### ISSUE TITLE
The local  API has changed 

### ISSUE THREAD

**Issue by tonybfox** [rithvikvibhu/GHLocalApi#39]
It looks like google have changed the api as I'm getting 403 Forbidden for anything other than /setup/eureka_info and that now return a public_key

**Comment by rithvikvibhu** [comment]
Thanks for letting me know! Will take a look at this in a few days. Hopefully, it's just an auth system that's easy to replicate.

**Comment by stboch** [comment]
I was digging in a bit to see if I can figure out what is going on... it seems that google has moved local communication to port 8443. I tried to do a little MitM but it seems that google is doing some application cert validation. (Either that or something else I did is wrong) They also changed it to require a client ssl for authentication.

**Comment by klimov-gett** [comment]
@stboch i think you're on point regarding application certificate:

i figure :8443 was google's port for SSL so i did a GET to https://192.168.1.219:8443/setup/bluetooth/status (since auto-connecting to bt speakers is what i do with this api)
as a result i received "Error: Peer certificate cannot be authenticated with given CA certificates"

after disabling certifcate validation i got 401, which reads "similar to 403, but in this case authorization is possible"

**Comment by cmorche** [comment]
What's odd for me is that this is still all working on 8008, and I'm even part of the preview program... This is the first time learning of the local API, so I guess I shouldn't bother spending too much time mucking about.

**Comment by ludeeus** [comment]
It looks like the change has rolled back, every endpoint I have now tested returns the expected result 🎉

**Comment by raddacle** [comment]
> It looks like the change has rolled back, every endpoint I have now tested returns the expected result 🎉

You're right! Though, I had to reboot my GH devices for it to revert

**Comment by jamieshaw** [comment]
I've been seeing this flip flop between available and unavailable for the past few days. Not entirely sure what's triggering it…

**Comment by Leatherface75** [comment]
I think they are preparing some sorts of authorization for higher security and maybe because of closing down NEST API and move it to the Cloud instead.

**Comment by rithvikvibhu** [comment]
So I finally got my hands on a GH and some updates.

First, the simple way of logging network traffic (with [PacketCapture](https://play.google.com/store/apps/details?id=app.greyshirts.sslcapture)) didn't work because it's a different port and other problems.
Tried setting up proxies with [burp](https://portswigger.net/burp) and [mitmproxy](https://mitmproxy.org), but it wasn't possible to forward other ports with [ProxyDroid](https://play.google.com/store/apps/details?id=org.proxydroid) (even with root). There is also SSL pinning, but [Frida](https://www.frida.re) was able to bypass that.
Finally, I set up a hotspot and started to capture with Wireshark, but during all this, my GH reverted back and I didn't have a chance to capture anything :/ Must have been because of one of the reboots. 
So yay that it's working right now, but auth may be back at any time. We'll just have to wait till then, I guess.

Simultaneously, I was going through the latest decompiled app and found some interesting bits. Most of the requests had an extra header `cast-local-authorization-token`. This is likely what we need to get. Unfortunately, the code is obfuscated and I couldn't find what this field should contain or how it is obtained. Will keep trying.

There are also more details on `/setup/eureka_info` than the docs, including a `local-authorization-token-hash` or something similar. Again, I didn't get to try this because it reverted back to not requiring any token.

Also, not sure since when, but there's a lot of Nest libraries now. Including Weave stuff (maybe https://openweave.io, haven't really checked it out yet).

That's it for now. If someone knows about the token or has any more info, feel free to comment here. I'm leaving this issue open.

**Comment by 6a61636b** [comment]
@rithvikvibhu All of my Google Home Mini in the UK are in a state where they return a 403, if I can help in debugging this let me know

```
curl -I http://bedroom-ghome.lan:8008/setup/eureka_info
HTTP/1.1 403 Forbidden
Content-Length:0
```

**Comment by rithvikvibhu** [comment]
Hey @6a61636b. That's interesting. `/setup/eureka_info` is the only one everyone could access when other endpoints returned 403.
Can you try `/setup/bluetooth/status`?

**Comment by 6a61636b** [comment]
```
curl -I http://bedroom-ghome.lan:8008/setup/bluetooth/status
HTTP/1.1 403 Forbidden
Content-Length:0
```

According to the Google Home app it is running System firmware version: 156414, Cast firmware version: 1.40.156414

**Comment by thorleifjaocbsen** [comment]
I can confirm the issue here in Norway on both a Chromecast Ultra and Chromecast 3 and the old Chromecast (not key fob). They all this morning returned 403. After a reboot they started working again.While they were at 403 I could not see any http data going through wireshark when using google home app. Everything went through 8443 and 8009.

**Comment by Sennevds** [comment]
> ```
> curl -I http://bedroom-ghome.lan:8008/setup/bluetooth/status
> HTTP/1.1 403 Forbidden
> Content-Length:0
> ```
> 
> According to the Google Home app it is running System firmware version: 156414, Cast firmware version: 1.40.156414

You need to use the small i.
So curl -i http://bedroom-ghome.lan:8008/setup/bluetooth/status

**Comment by thorleifjaocbsen** [comment]
> You need to use the small i.
> So curl -i http://bedroom-ghome.lan:8008/setup/bluetooth/status

Captial i is to get the "headers only" lowercase is to include the HTTP response. He just want to see the headers (403 forbidden) so capital i is perfect.

**Comment by Sennevds** [comment]
> > You need to use the small i.
> > So curl -i http://bedroom-ghome.lan:8008/setup/bluetooth/status
> 
> Captial i is to get the "headers only" lowercase is to include the HTTP response. He just want to see the headers (403 forbidden) so capital i is perfect.

That's true but still if I use:
`PC51:~$ curl -I http://192.168.0.5:8008/setup/bluetooth/status
 HTTP/1.1 405 Method Not Allowed
 Access-Control-Allow-Headers:Content-Type
 Cache-Control:no-cache
 Content-Length:0`

and if use:
`curl -i http://192.168.0.5:8008/setup/bluetooth/status
 HTTP/1.1 200 OK
 Access-Control-Allow-Headers:Content-Type
 Cache-Control:no-cache
 Content-Length:114
 Content-Type:application/json
 {"audio_mode":1,"connected_devices":[],"connecting_devices": [],"discovery_enabled":false,"scanning_enabled":false}`

**Comment by thorleifjaocbsen** [comment]
True, you need to add -X GET when using -I to specify method. Whereas -i somehow adds that manually. Wierd. 

Well spotted though

**Comment by alertedsnake** [comment]
Nothing weird about this.  The help explains:
```
-i, --include       Include protocol response headers in the output
-I, --head          Show document info only
```
A `GET request is the default with curl, and using `-i` will modify that request.  
`-I` means to make a HTTP `HEAD` request.  So the 405 response is not surprising if the endpoint only responds to `GET`

**Comment by rithvikvibhu** [comment]
@thorleifjaocbsen, how did you capture the packets? And where did you see the requests over ports 8009 and 8443? I'm asking this because normally http proxies won't be able to see them.

I'd love to take a look at that pcap file. Please send it to rithvikvibhu@gmail.com, if possible.

**Comment by thorleifjaocbsen** [comment]
I have a AP connected to a switch, then I have that port mirrored to my computer (lan adapter) and did dump all traffic going from and to the AP. My andorid is connected to WiFi the Chromecast is connected via ethernet. 

10.0.0.9 is my Android Device
10.0.0.8 is my Chromecast Device

I'm connecting using Google Home -> Configuring my Chromecast -> Changing some settings -> Rebooting it -> done pcap.

Filter: ip.addr == 10.0.0.8 or ip.addr == 10.0.0.9
Filter 2: (ip.src == 10.0.0.9 and ip.dst == 10.0.0.8) or (ip.dst == 10.0.0.9 and ip.src == 10.0.0.8)

Not sure if it helps. Sent you the file now :)

**Comment by Leatherface75** [comment]
What's the command to set an alarm/timer? I can only see commands for lists and delete.

**Comment by rithvikvibhu** [comment]
@thorleifjaocbsen Thanks for the file! There's not much to find there. 8443 uses TLS so no luck there. Any idea what those TCP packets on port 8009 are?

My GH is again back to sending 403 so I'll try some stuff with the android app.

**Comment by rithvikvibhu** [comment]
> 
> 
> What's the command to set an alarm/timer? I can only see commands for lists and delete.

@Leatherface75 There's no way to set new alarms/timers locally afaik.

**Comment by poma** [comment]
Mine are back to 403 too except the ones that have Preview Program turned on, looks like those have different firmware update schedule

**Comment by rithvikvibhu** [comment]
Hey everyone,

I got the API working over port 8443!
I'll keep this short. 

Turns out the extra header [I mentioned in the previous comment](https://github.com/rithvikvibhu/GHLocalApi/issues/39#issuecomment-509020118) was needed.
I kept digging in the app, but couldn't find what the value should be. Tried another method with frida and got the token value. Tried this value with postman, and it worked. `/setup/bluetooth/status` works and others should too, haven't tested.
As of now, I have no idea how often or when this token changes or expires, and from where this token is generated. After looking into this, I'll mostly post a guide or script to get it. For those curious, the token is part of Google's new Homegraph and you might be able to find it with root.

**Comment by rithvikvibhu** [comment]
Okay, created a gist with info:
https://gist.github.com/rithvikvibhu/1a0f4937af957ef6a78453e3be482c1f

Please try it out and see if it works.

**Comment by rithvikvibhu** [comment]
Also, I can't test a few cases. People with multiple rooted phones (with different Google accounts signed in), can you post the first and last few characters of the token? I'm curious if it's different for each home member.

**Comment by klimov-gett** [comment]
@rithvikvibhu thanks a bunch, tried the Frida method - works like a charm.
However, i think it's worth adding that you need (or at least i need) to pass '--insecure' option to curl for the request to go through. 
Reason for that being that the certificate installed on my Home Mini has no corresponding root CA / cert chain installed on my device's trust store and hence my Home is treated as a potential thread and curl aborts the request unless that option is passed.

**Comment by rithvikvibhu** [comment]
Hah yes, thanks @klimov-gett. I'll add that bit.

**Comment by Leatherface75** [comment]
That "public_key" is that different from that token you are talking about?

**Comment by rithvikvibhu** [comment]
@Leatherface75 That's the public key for a keypair for GH.
The token is different. It is used to authenticate requests made from the app to the GH.

**Comment by directman66** [comment]
Is it possible to generate a token without a rooted phone yet?

**Comment by Leatherface75** [comment]
This type om things normally need rooted devices.

**Comment by rithvikvibhu** [comment]
@directman66 not yet.

**Comment by Drakulix** [comment]
For anyone wanting a workaround for their home-assistant setup, if have written a small authentication proxy (you still need your token): https://github.com/Drakulix/ghlapi_proxy

Just point home-assistant to the IP of the proxy instead.


> As of now, I have no idea how often or when this token changes or expires, and from where this token is generated.

The token seems to change each time "your home" changes. I had to grab it again after:
- Adding a new device (new sonos speaker, so unrelated to my google home mini)
- Requesting a new sync (although nothing had actually changed) via: "Hey google, sync my devices".


> Okay, created a gist with info:
https://gist.github.com/rithvikvibhu/1a0f4937af957ef6a78453e3be482c1f

Your `decodeProtoFile.js `-script did not work for me at first, because it errored out on some devices of mine. I later noticed, those are my sonos speakers linked to my account, which do not have one of the indices of `val2` the script is expecting.

Just adding a try-catch block around the second lambda, correctly logged all tokens of all my devices: https://gist.github.com/Drakulix/e175ab49a416b298345d12eb555d6c24

**Comment by rithvikvibhu** [comment]
@Drakulix sweet!
I don't use home-assistant now, will keep yours in mind.

Just a few questions.

>The token seems to change each time "your home" changes.

Huh. Does the old token still work after you get the new one?
What is "Sync my devices" really meant to do?

Do all your devices have different tokens?
Maybe even non-GH devices (like Sonos) have tokens but at a different location.

I know that the file contains sensitive info to share, but can you try either formatting it with **[protoc](https://github.com/protocolbuffers/protobuf/releases)** (`cat homegraph.pb | protoc --decode_raw`) or opening it with a text editor? I just want to know if there's anything that resembles a token in the Sonos object.

**Comment by Leatherface75** [comment]
Sync devices is like it sounds.
Syncing devices so they have all informations including phones, tablets etc.
If you add a new lamp or anything else connected to Google Home you need to do that.

**Comment by jamieshaw** [comment]
@Drakulix Nice one on the proxy – works a treat. One issue though, and I'll open it on the repo is there isn't a simple answer? Is there a way to run this multiple times for multiple devices without some network wizardry spoofing an IP for each docker container as an endpoint for each Google Home device?  For the record; I have six… 🤔

(In short, there is no port option in the Google Home component within Home Assistant)

**Comment by OleksandrBerchenko** [comment]
My 5 cents:

According to this instruction I went to file system, copied the file and successfully extracted the token.  But the token DID NOT work.  Then I noticed that the file was modified on July 13.  I opened my Google Home application, waited for a few seconds and refreshed the file: it has been changed.  Copied it again and extracted the token: it was different and actually worked.

Two conclusions:

1. I didn't do anything special with my devices since July 13, but the token changed.
2. The phone don't get a new token automatically, unless you manually run Google Home application.

**Comment by Drakulix** [comment]
> > The token seems to change each time "your home" changes.
> 
> Huh. Does the old token still work after you get the new one?
> What is "Sync my devices" really meant to do?

No, the old token does not work anymore, that is how I noticed that.
"Sync my devices" starts a new request to your providers to register all connected devices with home-graph. This is sometimes necessary especially with home-assistant if my google home is not picking up new devices correctly.

> Do all your devices have different tokens?
> Maybe even non-GH devices (like Sonos) have tokens but at a different location.

All my other devices (two Android TV Boxes) do have tokens as well, although I am not aware of any http-intefaces. The Sonos ones might have some tokens as well, but they do not support the local-api, so I do not really care.

> I know that the file contains sensitive info to share, but can you try either formatting it with **[protoc](https://github.com/protocolbuffers/protobuf/releases)** (`cat homegraph.pb | protoc --decode_raw`) or opening it with a text editor? I just want to know if there's anything that resembles a token in the Sonos object.

I do not see anything interesting, but here is one of my sonos speakers (with removed id-like structures and names):

<details><summary>Sonos-object protobuf decoded</summary><pre>
  7 {
    1 {
      1: "(uuid)"
      2 {
        1: "sonos-(five digits+letters)"
        2: "RINCON_(17 digits+letters)"
      }
    }
    4: "(device name)"
    6: "action.devices.types.SPEAKER"
    7: "action.devices.traits.MediaInitiation"
    7: "action.devices.traits.TransportControl"
    7: "action.devices.traits.Volume"
    7: "action.devices.traits.MediaState"
    7: "action.devices.traits.Assistant"
    7: "action.devices.traits.InputSelector"
    17 {
      1: "Sonos"
    }
    18 {
      2: "(room name)"
      3: "Sonos"
    }
    19: (13 digits)
    20 {
      1 {
        1: "availableInputs"
        2 {
          6 {
            1 {
              5 {
                1 {
                  1: "key"
                  2 {
                    3: "antenna"
                  }
                }
              }
            }
          }
        }
      }
      1 {
        1: "canReportMediaState"
        2 {
          4: 1
        }
      }
      1 {
        1: "commandOnlyOnOff"
        2 {
          4: 0
        }
      }
      1 {
        1: "commandOnlyVolume"
        2 {
          4: 0
        }
      }
      1 {
        1: "levelStepSize"
        2 {
          2: 0x4014000000000000
        }
      }
      1 {
        1: "mediaInitiationAvailableProviders"
        2 {
          6 {
            1 {
              3 {
                8: 0x6f69646152756363
              }
            }
            1 {
              3: "Amazon Music"
            }
            1 {
              3: "AMPed"
            }
            1 {
              3: "Anghami"
            }
            1 {
              3: "Audible"
            }
            1 {
              3: "Bandcamp"
            }
            1 {
              3: "Classical Archives"
            }
            1 {
              3: "Deezer"
            }
            1 {
              3: "Gaana"
            }
            1 {
              3: "Google Play Music"
            }
            1 {
              3: "iHeartRadio"
            }
            1 {
              3: "JUKE"
            }
            1 {
              3: "KKBOX"
            }
            1 {
              3: "MeloMe"
            }
            1 {
              3: "Mixcloud"
            }
            1 {
              3: "My Cloud Home"
            }
            1 {
              3: "NhacCuaTui"
            }
            1 {
              3: "Pandora"
            }
            1 {
              3: "Plex"
            }
            1 {
              3 {
                10: 114
                13: 0x63696e6f6870656d
              }
            }
            1 {
              3: "Qobuz"
            }
            1 {
              3: "QQ\351\237\263\344\271\220"
            }
            1 {
              3: "Radio Javan"
            }
            1 {
              3: "RadioPlay"
            }
            1 {
              3: "rova"
            }
            1 {
              3: "Saavn"
            }
            1 {
              3: "SoundCloud"
            }
            1 {
              3: "Stingray Music"
            }
            1 {
              3: "TIDAL"
            }
            1 {
              3: "TuneIn"
            }
            1 {
              3: "WEYV"
            }
            1 {
              3: "YouTube Music"
            }
            1 {
              3: "\345\207\244\345\207\260FM"
            }
            1 {
              3: "\345\215\203\345\215\203\351\237\263\344\271\220(\345\216\237\347\231\276\345\272\246\351\237\263\344\271\220)"
            }
            1 {
              3: "\345\226\234\351\251\254\346\213\211\351\233\205FM"
            }
            1 {
              3: "\347\275\221\346\230\223\344\272\221\351\237\263\344\271\220"
            }
            1 {
              3: "Spotify"
            }
          }
        }
      }
      1 {
        1: "mediaInitiationByDescription"
        2 {
          4: 1
        }
      }
      1 {
        1: "mediaInitiationCanPlayAudio"
        2 {
          4: 1
        }
      }
      1 {
        1: "needsAssistantUnlink"
        2 {
          4: 1
        }
      }
      1 {
        1: "onDemandAssistantDiscoverability"
        2 {
          4: 1
        }
      }
      1 {
        1: "orderedInputs"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportActionState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportActivityState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportApplicationState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportIsPlayingMediaState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportPlaybackState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportTimeSyncState"
        2 {
          4: 1
        }
      }
      1 {
        1: "supportedMediaTypes"
        2 {
          6 {
            1 {
              3: "Language"
            }
            1 {
              3: "MusicRecording"
            }
            1 {
              3: "MusicAlbum"
            }
            1 {
              3: "MusicGroup"
            }
            1 {
              3: "MusicPlaylist"
            }
            1 {
              3: "Person"
            }
            1 {
              3: "RadioChannel"
            }
            1 {
              3: "RadioBroadcastService"
            }
            1 {
              3: "Thing"
            }
          }
        }
      }
      1 {
        1: "transportControlSupportedCommands"
        2 {
          6 {
            1 {
              3: "NEXT"
            }
            1 {
              3: "PREVIOUS"
            }
            1 {
              3: "PAUSE"
            }
            1 {
              3: "STOP"
            }
            1 {
              3: "RESUME"
            }
            1 {
              3: "SEEK_RELATIVE"
            }
          }
        }
      }
      1 {
        1: "volumeCanMuteAndUnmute"
        2 {
          4: 1
        }
      }
      1 {
        1: "volumeDefault"
        2 {
          2: 0x4024000000000000
        }
      }
      1 {
        1: "volumeMaxLevel"
        2 {
          2: 0x4059000000000000
        }
      }
    }
    26 {
      1: "(google account mail)"
    }
    34 {
      1: 1
    }
    16: "Sonos"
  }
</pre></details>



> @Drakulix Nice one on the proxy – works a treat. One issue though, and I'll open it on the repo is there isn't a simple answer? Is there a way to run this multiple times for multiple devices without some network wizardry spoofing an IP for each docker container as an endpoint for each Google Home device? For the record; I have six… thinking
> 

Yes there is, if your docker-host is connected via ethernet to your home network, you may setup a macvlan (or ipvlan) network via docker and assign that to all your proxy containers. Every container should then get its own mac or ip address.

> (In short, there is no port option in the Google Home component within Home Assistant)

Yeah I noticed that too, total bummer.

> 2. The phone don't get a new token automatically, unless you manually run Google Home application.

Can confirm that, I always have to open the google home app on my rooted device before I can extract new tokens. This makes automatic extraction rather annoying.

**Comment by Leatherface75** [comment]
Maybe something that can upload that file once a day or something to server there it extracts and updates the token?

**Comment by Leatherface75** [comment]
I am getting this error after it finds first google home token.

/tmp/decodeProtoFile.js:24
                var token = getObjByKey(val2['7'], '28')['28'];
                                                        ^

TypeError: Cannot read property '28' of undefined
    at /tmp/decodeProtoFile.js:24:57
    at Array.forEach (<anonymous>)
    at /tmp/decodeProtoFile.js:21:18
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/tmp/decodeProtoFile.js:19:6)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:837:10)

**Comment by Leatherface75** [comment]
If i change this line
var token = getObjByKey(val2['7'], '28')['28'];
to
var token = getObjByKey(val2['7'], '28');

then i am able to get tokens for all Google Home devices but it shows lot of other things from Google Assistant too like lamps, consoles, tvboxes, androidtv etc.

Device:  undefined, Taklampa Vardagsrum
Token:   undefined
-----
-----
Device:  Google Home Mini, Högtalare Kök
Token:   { '28':
   'TOKEN-REMOVED'}
-----
-----
Device:  undefined, [object Object],[object Object],[object Object],[object Object]
Token:   undefined
-----
-----
Device:  Google Home Mini, Högtalare Sovrum
Token:   { '28':
   'TOKEN'-REMOVED }
-----
-----
Device:  undefined, MiBox3 Sovrum
Token:   undefined
-----
-----
Device:  undefined, RPi3 Sovrum
Token:   undefined

**Comment by tatunts** [comment]
> A better solution would be something that can communicate with Google Home API and get tokens from there. https://developers.google.com/actions/smarthome/reference/rest/

Hello sir! I was thinking about that as well. Wouldn't be easier to connect to the device using those API over the internet? Not sure if there is an API available for that or how to identify the correct Google Home Mini device you need to get the alarm information from.

I am a Home Assistant user and already have configured a test app needed to expose my devices to the GH API and controle them.

Sorry, if can not help you guys solve this problems since I lack the proper coding skills...

Thanks!

**Comment by Leatherface75** [comment]
That home_graph file containt that information and tha API is in that link.

**Comment by Leatherface75** [comment]
As a workaround for you guys with more than 1 google home and Linux or similiar i modified the proxy code above and added a number and compiled in 3 versions.
Then i did this

ip addr add 192.168.0.120/24 dev enp2s0
ip addr add 192.168.0.121/24 dev enp2s0
ip addr add 192.168.0.122/24 dev enp2s0
export SOURCE_IP_1=192.168.0.15
export SOURCE_IP_2=192.168.0.30
export SOURCE_IP_3=192.168.0.70
export LISTEN_IP_1=192.168.0.120
export LISTEN_IP_2=192.168.0.121
export LISTEN_IP_3=192.168.0.122
export TOKEN_1=TOKEN_FOR_GOOGLE_HOME_1
export TOKEN_2=TOKEN_FOR_GOOGLE_HOME_2
export TOKEN_3=TOKEN_FOR_GOOGLE_HOME_3
ghlapi_proxy_1 &
ghlapi_proxy_2 &
ghlapi_proxy_3 &

Then change your configuration.yaml file if you're using Home Assistant with the new IP's
Change IP's and name of network card for your setup.

**Comment by Leatherface75** [comment]
and it didn't took long time and tokens was changed so this method doesn't work if you don't want to update your token once or a few time a day.

**Comment by KapJI** [comment]
@rithvikvibhu did you try to figure out how app is obtaining these tokens? As their lifespan is quite short we need another way than getting token from a rooted device.

**Comment by Leatherface75** [comment]
Home assistant for example is communicating and have their own token for Google assistant.
Shouldnt it be possible to get information with update addon there?

**Comment by Leatherface75** [comment]
Installed Android on a virtual machine with Home app open and then a crontab that grabs that file once in an hour. Let's see how that works been working for almost one day now.

**Comment by jamieshaw** [comment]
Can you share which virtual machine you used? I've been looking at spinning up an Android VM but couldn't find anything of worth after browsing (for a very short period of time).

**Comment by Leatherface75** [comment]
I am Using ESXi Server but any virtualization software should work fine.
Here you can find Android images https://www.android-x86.org

But as an update this is no optimal solution either i have to restart Google Home app sometimes because it stop updating and when that happens it doesn't work in Google's own app either.

**Comment by jamieshaw** [comment]
Thanks, I'm also using ESXi 😉

(Even if not automated, saves me a job of having to source an Android phone purely for this task)

**Comment by Leatherface75** [comment]
I'm going to play a little with Tasker for that restarting of Home app.

**Comment by ronmichel** [comment]
Is there something like a `agentUserId` in the homegraph file? According to [this](https://developers.google.com/actions/smarthome/reference/rpc/google.home.graph.v1)
Homegraph API allows storing and querying **first-party** and third-party devices (I assume a Google Home would be a first-party device). But for the status [query ](https://developers.google.com/actions/smarthome/reference/rest/v1/devices/query) there is always a Third-party user ID needed (and the device ID) with no mention of how to query a first-party device.

**Comment by thorleifjaocbsen** [comment]
May I jump in here, our solution uses Chromecast's only. Not google home, We have a few buildings which have Chromecasts in conference rooms. Where we control them by using this google home API (reboot, set name, start app e.t.c). Now that seems like it might be impossible, do this device you get token from have to be the device that set the chromecasts up? As we used the API to configure them aswell, never used google home. We connected them to ethernet then did all config via API.

How would this affect me ? Can I still get a common token for all devices in all buildings without connecting a phone to them?

**Comment by Leatherface75** [comment]
It just needs to be in same google account there you have registered all devices.
That homegraph file you can get with root or android in virtual machine have all information for all devices you have in your Google Home app.
Even all things exported from Home Assistant is there.

**Comment by thorleifjaocbsen** [comment]
The problem for us is that none of the devices is set up with a Google account. All is configured via the API. So we basically loose control now. That was boring. It just seems to have arrived to chromecast yesterday as our software cannot control them anymore :(

**Comment by Leatherface75** [comment]
As long your software are connected to same google account your chromecasts are registered with it should not change anything for the others.

**Comment by thorleifjaocbsen** [comment]
Ah, our software just sends the setup/eureka_info to change name and the reboot command. So neither is ever registered to an Google account weirdly it still gets blocked

**Comment by ronmichel** [comment]
Did anyone check if the token also changes if there is no internet connection but the devices only connected to the local network?

**Comment by Leatherface75** [comment]
> Ah, our software just sends the setup/eureka_info to change name and the reboot command. So neither is ever registered to an Google account weirdly it still gets blocked

You need that token that is explained above.

**Comment by Leatherface75** [comment]
Found a better way to restart Google Home than Tasker

Adding this to crontab script solves that

adb shell ps|grep chromecast|awk '{print $2}'|xargs adb shell kill
adb shell monkey -p com.google.android.apps.chromecast.app -c android.intent.category.LAUNCHER 1

**Comment by jiauka** [comment]
I have successfully controlled the ChromeCast by connecting it tru ethernet -not wifi- using an USB to Ethernet dongle and and "Y" (microUSB female, microUSB male and USB A female) cable.

Then I'm able to use all the previous 8008 port commands by using the 8443 port and sending my OWN Authorization field, i.e.

`curl -Lv -k -X POST -H "Content-Type: application/json" -H "Authorization: roomy.cast" -d '{"name": "ITWORKS"}' https://192.168.67.210:8443/setup/set_eureka_info
` 

I can issue all the commands as long as no phone/tablet gets connected to the Chromecast, in my application I do avoid it setting iptables filters at the router where the Chromecast is connected.

**Comment by rithvikvibhu** [comment]
@jiauka to be clear, you sent a _random_ `Authorization` token and all endpoints work over port 8443?

**Comment by jiauka** [comment]
Yes, a random Auth works as long as:
1. Device comes from a Factory Reset
2. It is connected via Ethernet
3. None phone/tablet has edited it tru Google Home App.

It is also possible to use port 8008 but only ONCE, following calls end up with 403.

Probably it can connected to Wifi via curls to the setup/set_eureka_info stuff.

In my application I blick ports 8443 and 8008 and I'm happy

**Comment by Leatherface75** [comment]
That makes the Google Home unusable for other things so doesn't sound like a good idea for me.
You can't set timer/alarm or anything because that will call HomeGraph API.

**Comment by ronmichel** [comment]
@rithvikvibhu 
Did you see [this](https://github.com/actions-on-google/smart-home-dashboard) (can be used to retrieve the data directly from homegraph)

**Comment by jamieshaw** [comment]
Has the Homegraph response changed? As of this morning, my kitchen Google Home mini (1 of 6 throughout the house) is yielding a different response from my automation for pulling out the authorisation token; normally traversing through the tree, one object has key '28' containing the token as a string.

This morning, by Kitchen mini has an array for this position, of which key `[0][13]` has a similar string, but this is failing to authenticate against the mini.  The other speakers, are presently unaffected…

**Comment by mhmorales** [comment]
@rithvikvibhu Can you help me?

I have the following errror

![image](https://user-images.githubusercontent.com/8845715/64025790-dcc18d00-cb13-11e9-99d2-cdb1341e77a1.png)

Regards

**Comment by rithvikvibhu** [comment]
@Leatherface75 the docs over at developers.google.com/actions/ are for allowing users to control your services through Google Home. That's like adding 3rd party service (ex: Philips Hue, etc.) and controlling them through GH app or Google Assistant. It is not related to GH's local API used by the mobile app to communicate with the device.

@ronmichel Same as above. That is only for 3rd parties to link and get info about their own devices.

**Comment by rithvikvibhu** [comment]
@jamieshaw and @mhmorales That's interesting. This is expected as that's how protobuf works. The structure keeps changing as the generated client libraries changed.
I'm thinking of regexing for the tokens instead of navigating with hardcoded indexes. Will update here later.
This was only a temp measure anyway until there's a way to generate or get tokens without the app or an android vm.

**Comment by Leatherface75** [comment]
> @Leatherface75 the docs over at developers.google.com/actions/ are for allowing users to control your services through Google Home. That's like adding 3rd party service (ex: Philips Hue, etc.) and controlling them through GH app or Google Assistant. It is not related to GH's local API used by the mobile app to communicate with the device.
> 
> @ronmichel Same as above. That is only for 3rd parties to link and get info about their own devices.

I know but i am thinking just for getting information about the devices.
That homegraph file have all information about lamps, all integrations i have done in Home Assistant etc everything is there.
When you have that information you can control the device locally again as before.
All integrations i can se does some sort of sync request to that API.

**Comment by Drakulix** [comment]
I don't know a better place to put this, so sorry for hi-jacking again.
For everyone looking for a replacement for home-assistant, I have brewed all of this into a custom_component: https://github.com/Drakulix/googlehome

Just make sure your HA-instance has access to an adb server with a connected rooted device and it will automatically restart the Google Home App, pull the protocol file, decode it and integrate everything as normal.

**Comment by Leatherface75** [comment]
Thanks for this does it work for multiple google homes?
My solution i have now works perfectly now with 3 compiled proxys, crontab that updates tokens from my virtual machine. But in my case i have to restart and do some keyevents after before it creates an updated homegraph file. Does this take care of that?

**Comment by Drakulix** [comment]
> Thanks for this does it work for multiple google homes?
> My solution i have now works perfectly now with 3 compiled proxys, crontab that updates tokens from my virtual machine with Android. But in my case i have to restart and do some keyevents after before it creates an updated homegraph file. Seems it doesn't always create that file after a restart only. After adding "adb shell input keyevent 4" in my crontab after restart it seems to always do that.
> Does this take care of that?
> Tested your component now and i am only getting error on adb commands in config.

Yes, this should just work fine for multiple google homes, because it just fetches the tokens for all of them and then the specific clients try to grab their token.

What kind of keyevents are you sending?
I can easily integrate them: https://github.com/Drakulix/googlehome/blob/master/custom_components/googlehome/adb_helper.py#L22

I have noticed this as well, but the app will fetch the new file eventually. Because my integration fetches new tokens as soon, as a request fails, this was sufficient for now in my case. But of course this can be improved, just open an issue on the repository and we can continue the discussion there.

**Comment by Leatherface75** [comment]
`adb shell input keyevent 4`

This keyevents a while after restarting have worked for a while for me.

**Comment by Leatherface75** [comment]
You seem to have an older version than me i am on 1.42.171861 but i am in beta.

**Comment by fmieres** [comment]
> You seem to have an older version than me i am on 1.42.171861 but i am in beta.

ive got  a few devices in   "cast_build_revision": "1.40.156414", and they do need auth token to be requested. i thought they may have rolled back this issue on a later release, but since you have a versiion ahead, it doesnt seem the case anymore :/

To make it worse, i dont even have a token in the protofile in devices connected to 1.42.17

**Comment by Leatherface75** [comment]
I don't know if your method works or not i haven't tested since my setup i have now works for me.

**Comment by magicse** [comment]
list of endpoint of api
/setup/accept_tos
/setup/configured_networks
/setup/dummy_impl
/setup/eureka_info
/setup/forget_wifi
/setup/get_app_device_id
/setup/get_local_authorization_token
/setup/get_system_property
/setup/offer
/setup/play_sound
/setup/reboot
/setup/resume_cast
/setup/save_wifi
/setup/scan_results
/setup/scan_wifi
/setup/send_log_report
/setup/set_eureka_info
/setup/set_system_property
/setup/set_temporary_update
/setup/start_nat
/setup/supported_locales
/setup/supported_timezones
/setup/suspend_cast
/setup/test_internet_download_speed
/setup/test_udp_throughput
/setup/trigger_weave_pairing

**Comment by magicse** [comment]
May be some additional names of endpoints
ble_setup
http_api_get_log_report
http_api_get_log_report_default
http_api_handler
http_api_test_tcp_throughput
mojo_api_proxy
mojo_eureka_info_proxy
mojo_setup_manager
send_log_report_request_handler
send_log_report_request_handler_fake
send_log_report_request_handler_linux
setup_constants
setup_manager_impl
setup_namespace_handler
setup_state

**Comment by magicse** [comment]
/setup/assistant/alarms
/setup/assistant/alarms_delete
/setup/assistant/alarms_options
/setup/assistant/alarms_volume
/setup/assistant/can_enroll_status
/setup/assistant/light_eq
/setup/assistant/set_led_brightness
/setup/assistant/set_mute_indicator_functional
/setup/assistant/set_network_status_led_pattern
/setup/assistant/set_night_mode_params
/setup/assistant/set_tts_volume
/setup/assistant/show_led

**Comment by Leatherface75** [comment]
interesting some things maybe could improve local usage of things.

**Comment by magicse** [comment]
May be this help
Link to GH sources for different devices with chromecast and assistant
[chromecast oss](https://drive.google.com/drive/folders/0B3j4zj2IQp7MZkplRzRvcERtaU0l)

**Comment by rithvikvibhu** [comment]
@magicse thanks for the ton of new stuff! Where did you get these? Because they aren't used by the app. Some of them return a 404. As @Leatherface75 said, it could be device-dependent, but I still don't know how it's not in the app.
Do you have any more info about these (what body to send, etc.)? I'm particularly interested in `get_local_authorization_token` which is a pain to extract right now and would simplify things a lot.

**Comment by Leatherface75** [comment]
Yeah would be nice to drop virtual machine with Android just for the token but for now that still works.

**Comment by XoMEX** [comment]
You can find this google drive link by googling "chromecast oss". You will find this help article: https://support.google.com/chromecastbuiltin/answer/6121012?hl=en which in turn links to the code.
Unfortunately, I don't see any (more) helpful information in that article.

As for the archive contents. It seems like the interesting parts are not available as source, but only as object files (I only took a look at `chromecast_oss/1.46/google_home_3`; Did not test if other downloads may have more/less info). Fortunately for us, they contain debug info and are not stripped.

Taking a look at `chromecast::ApiGetLocalAuthorizationToken::OnApiRequest` (file `chromecast_oss/chromium/src/out_chromecast_biggie/release/obj/chromecast/internal/setup/setup/api_get_local_authorization_token.o`) we can see that it uses a `LocalAuthorizationTokenManager`.

First, we can briefly look at the tokens (see `chromium/src/out_chromecast_biggie/release/gen/chromecast/internal/certificate/proto/local_authorization_token.pb.cc`). They contain the `cloud_device_id` and a `sequence_number`.

This manager class is located in `chromecast_oss/chromium/src/out_chromecast_biggie/release/obj/chromecast/internal/certificate/certificate/local_authorization_token_manager.o`.
It has several functions; Let's start at the constructor. The main interest seems to be `LoadSequenceNumber` and `IncrementLocalAuthorizationToken`. The first function indicates that the sequence number is persistent across reboots. The second function simply increases this counter and calls `GenerateLocalAuthorizationToken`. This creates a LocalAuthorizationToken object and sets the cloud device id as well as the sequence number. This then gets serialized and encrypted (I think AES256GCM). Then the OnTokenEncrypted function is called (seemingly indirect, possibly asynchronous). This calls all observers of the manager. I assume one observer will call `LATUploaded` which then enforces the auth token requirement.
The name of this lat function indicates to me that the token is sent to the cloud, from where it is sent to the app.

(Note: This information was gathered using ghidra. This was cpp code, which results in rather ugly decompiled code. Thus, I had to make some educated guesses.)

My Ideas on ways to get the token are:
- Somehow get ahold of the key and generate your own tokens (unlikely: I *assume* the key is generated using some unique hardware id or some key established with the cloud during setup)
- Find out which API is used to get the key from the cloud (In the decompiled app there is a call to a function with the parameters `"google.internal.home.foyer.v1.StructuresService", "GetHomeGraph"`, this *could* be something, but the decompiled app is very hard to work with)
- Find out that I made some wrong conclusions during my reversing and that there is another way

**Comment by rithvikvibhu** [comment]
@XoMEX it's great you've been able to find all this! I had downloaded Google Home 1 (pineapple) and saw similar files. So they probably won't have included the sources in other builds too.

Since `api_get_local_authorization_token.o` file handles the token generation and upload, not all files in the setup folder correspond to an API endpoint.

The whole process makes sense. Sending the token to the clients through the cloud is pretty smart as the clients must sign in with the correct Google account first.

>- Somehow get ahold of the key and generate your own tokens

Yes, interfering with the device itself isn't something that everyone would want to do.

>- Find out which API is used to get the key from the cloud

I've been trying this, but as you said, obfuscated code is a pain to read. I did come across `foyer` but dismissed it thinking it's related to Google WiFi. So I went back to look at this again, and there are a lot of functions tied to this.

<details>
<summary>List of foyer related methods</summary>

```
"google.internal.home.foyer.v1.AppDataService", "GetAppConfigurationData"
"google.internal.home.foyer.v1.AppDataService", "GetAssistantLearn"
"google.internal.home.foyer.v1.AppDataService", "GetAssistantLearn"
"google.internal.home.foyer.v1.AppDataService", "GetSummaryList"
"google.internal.home.foyer.v1.AssistCardsService", "Dismiss"
"google.internal.home.foyer.v1.AssistCardsService", "List"
"google.internal.home.foyer.v1.CallsService", "DisableDuoCalls"
"google.internal.home.foyer.v1.CallsService", "EnableDuoCalls"
"google.internal.home.foyer.v1.CallsService", "GetDuoSettings"
"google.internal.home.foyer.v1.CallsService", "UpdateDuoSettings"
"google.internal.home.foyer.v1.CameraService", "GetCameraMetadata"
"google.internal.home.foyer.v1.CameraService", "UpdateCameraProperty"
"google.internal.home.foyer.v1.ClientOliveTokenService", "GetOliveToken"
"google.internal.home.foyer.v1.DynamicRpcInvocationProxyService", "InvokeRpc"
"google.internal.home.foyer.v1.EnergyService", "GetThermostatStatus"
"google.internal.home.foyer.v1.FeedManagementService"
"google.internal.home.foyer.v1.HistoryManagementService", "GetFilters"
"google.internal.home.foyer.v1.HistoryManagementService", "GetHistory"
"google.internal.home.foyer.v1.HomeControlService", "ExecuteRoutineOnDevice"
"google.internal.home.foyer.v1.HomeControlService", "GetAssistantRoutines"
"google.internal.home.foyer.v1.HomeControlService", "InitiateMediaPlayback"
"google.internal.home.foyer.v1.HomeDevicesService", "CheckDeviceReadyState"
"google.internal.home.foyer.v1.HomeDevicesService", "CreateDevice"
"google.internal.home.foyer.v1.HomeDevicesService", "DeleteDevice"
"google.internal.home.foyer.v1.HomeDevicesService", "GetAssistantDeviceSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "GetBootstrappedDevices"
"google.internal.home.foyer.v1.HomeDevicesService", "GetBootstrappedDevices"
"google.internal.home.foyer.v1.HomeDevicesService", "GetDefaultOutputDevices"
"google.internal.home.foyer.v1.HomeDevicesService", "GetDevicesAssociations"
"google.internal.home.foyer.v1.HomeDevicesService", "GetDeviceSupervisionSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "GetRoutineAlarmSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "GetSensingSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "SendEnterWeavePairingModeCommand"
"google.internal.home.foyer.v1.HomeDevicesService", "SetDefaultOutputDevices"
"google.internal.home.foyer.v1.HomeDevicesService", "SyncDevices"
"google.internal.home.foyer.v1.HomeDevicesService", "UpdateAssistantDeviceSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "UpdateDeviceSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "UpdateDeviceSupervisionSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "UpdateRoutineAlarmSettings"
"google.internal.home.foyer.v1.HomeDevicesService", "UpdateSensingSettings"
"google.internal.home.foyer.v1.OffersService", "ListOffers"
"google.internal.home.foyer.v1.SetupService", "GetDeviceDigitalUserGuide"
"google.internal.home.foyer.v1.SetupService", "GetHighlightedApplications"
"google.internal.home.foyer.v1.SpacesService", "BatchModifySpacesDevices"
"google.internal.home.foyer.v1.SpacesService", "CreateSpace"
"google.internal.home.foyer.v1.SpacesService", "CreateSpaceAndMoveDevices"
"google.internal.home.foyer.v1.SpacesService", "DeleteSpace"
"google.internal.home.foyer.v1.SpacesService", "UpdateSpace"
"google.internal.home.foyer.v1.StructuresService", "AcceptManagerInvite"
"google.internal.home.foyer.v1.StructuresService", "BatchModifyStructuresDevices"
"google.internal.home.foyer.v1.StructuresService", "ConvertApplicantToInvitee"
"google.internal.home.foyer.v1.StructuresService", "CreateApplicationForInvite"
"google.internal.home.foyer.v1.StructuresService", "CreateManagerInvite"
"google.internal.home.foyer.v1.StructuresService", "CreateNestStructure"
"google.internal.home.foyer.v1.StructuresService", "CreateStructure"
"google.internal.home.foyer.v1.StructuresService", "DeleteManagerInvite"
"google.internal.home.foyer.v1.StructuresService", "DeleteStructure"
"google.internal.home.foyer.v1.StructuresService", "GetHomeGraph"
"google.internal.home.foyer.v1.StructuresService", "GetInvitedStructure"
"google.internal.home.foyer.v1.StructuresService", "GetUserAddress"
"google.internal.home.foyer.v1.StructuresService", "GetWeavePairingInfo"
"google.internal.home.foyer.v1.StructuresService", "GetWorksWithNestState"
"google.internal.home.foyer.v1.StructuresService", "RejectApplicationForInvite"
"google.internal.home.foyer.v1.StructuresService", "RejectManagerInvite"
"google.internal.home.foyer.v1.StructuresService", "RemoveStructureManager"
"google.internal.home.foyer.v1.StructuresService", "ResendManagerInviteEmail"
"google.internal.home.foyer.v1.StructuresService", "UpdatePairingConfigurationDone"
"google.internal.home.foyer.v1.StructuresService", "UpdateStructureV2"
"google.internal.home.foyer.v1.UsersService", "AppSubscriptionUpdate"
"google.internal.home.foyer.v1.UsersService", "CreateNestAccount"
"google.internal.home.foyer.v1.UsersService", "GetNestAccountLinkState"
"google.internal.home.foyer.v1.UsersService", "GetUserPreferences"
"google.internal.home.foyer.v1.UsersService", "GetWwnAccountLinkState"
"google.internal.home.foyer.v1.UsersService", "RegisterFcmMessaging"
"google.internal.home.foyer.v1.UsersService", "SetUserPreferences"
"google.internal.home.foyer.v1.UsersService", "UnregisterFcmMessaging"
```
</details>

Looking further, some other things I noticed:
- this seems to be using gRPC, but with the obfuscated code, it's hard to get the specifics.
- A scope(?) seems to be added whenever these requests are made: `oauth2:https://www.googleapis.com/auth/homegraph`
- There is a Bearer token (Google-style access tokens, `ya29.xxx`) used somewhere, not sure if it is part of grpc calls or other requests.
- Many (including the `getHomeGraph` method) are `UNARY` calls, [more info here](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition)
- There is a function (no point in naming it, as obfuscated names would change every time) that joins the "fullServiceName" (first col in the above list) and "methodName" (second col), separated by a `/`. The combined string is also visible below.
- The URL all these are made to seems to be either `googlehomefoyer-pa.googleapis.com` or `https://clients3.google.com` (likely the first one)
- I logged all runtime strings and a few other things and this supports the other points. Interesting strings below.

<details>
<summary>Runtime strings and other stuff</summary>

```
{'type': 'send', 'payload': '[0] /googlehomefoyer-pa.googleapis.com:443'}
{'type': 'send', 'payload': "[0] Channel for 'directaddress:///googlehomefoyer-pa.googleapis.com:443'"}
{'type': 'send', 'payload': '[0] grpc-default-executor-0'}
{'type': 'send', 'payload': '[0] yij{logId=1, target=directaddress:///googlehomefoyer-pa.googleapis.com:443}'}
{'type': 'send', 'payload': '[0] com.google.android.apps.chromecast.app/21450110 (Linux; U; Android 5.0; en_US; Google Nexus 4_1; Build/LRX21M; Cronet/81.0.4044.96)'}
{'type': 'send', 'payload': '[0] /people-pa.googleapis.com:443'}
{'type': 'send', 'payload': "[0] Channel for 'directaddress:///people-pa.googleapis.com:443'"}
{'type': 'send', 'payload': '[0] yij{logId=3, target=directaddress:///people-pa.googleapis.com:443}'}
{'type': 'send', 'payload': '[0] Must be called on GoogleApiHandler thread, but got GoogleApiHandler.'}
{'type': 'send', 'payload': '[0] 1-https://clients3.google.com/cast/orchestration/linkeddevices?rt=b'}
{'type': 'send', 'payload': '[0] Bearer ya29.a0AfH6*****'}
[*] Created new URL with value: https://clients3.google.com/cast/orchestration/linkeddevices?rt=b
{'type': 'send', 'payload': '[0] google.internal.home.foyer.v1.StructuresService/GetHomeGraph'}
{'type': 'send', 'payload': '[0] Sent.google.internal.home.foyer.v1.StructuresService.GetHomeGraph'}
{'type': 'send', 'payload': '[0] grpc-timer-0'}
{'type': 'send', 'payload': '[0] [[googlehomefoyer-pa.googleapis.com:443]/{}]'}
{'type': 'send', 'payload': '[0] Subchannel for [[[googlehomefoyer-pa.googleapis.com:443]/{}]]'}
{'type': 'send', 'payload': '[0] Cannot transition out of SHUTDOWN to CONNECTING'}
{'type': 'send', 'payload': '[0] grpc-java-cronet/1.25.0-SNAPSHOT'}
{'type': 'send', 'payload': '[0] Cannot transition out of SHUTDOWN to READY'}
{'type': 'send', 'payload': '[0] grpc-default-executor-0'}
{'type': 'send', 'payload': '[0] 1-https://clients3.google.com/cast/discover/config'}
{'type': 'send', 'payload': '[0] 1-https://clients3.google.com/cast/discover/config'}
{'type': 'send', 'payload': '[0] Bearer ya29.a0AfH6*****'}
[*] Created new URL with value: https://clients3.google.com/cast/discover/config
{'type': 'send', 'payload': '[0] https://googlehomefoyer-pa.googleapis.com:443/google.internal.home.foyer.v1.StructuresService/GetHomeGraph'}
{'type': 'send', 'payload': '[0] https://clients3.google.com/'}
```
</details>

I don't know much about grpc and as of now, have no way of making these calls. Does anyone here have any experience with https://grpc.io?

**Comment by magicse** [comment]
additional info [https://www.researchgate.net/figure/Google-Home-Mini-network-setup-for-device-mobile-app-and-network-data-collection_fig1_338593589](url)

**Comment by magicse** [comment]
Additional info [](https://www.researchgate.net/figure/Google-Home-Mini-network-setup-for-device-mobile-app-and-network-data-collection_fig1_338593589)

**Comment by magicse** [comment]
additional info https://www.researchgate.net/figure/Google-Home-Mini-network-setup-for-device-mobile-app-and-network-data-collection_fig1_338593589

**Comment by magicse** [comment]
https://documenter.getpostman.com/view/7490211/SzzdD1pF?version=latest

**Comment by rithvikvibhu** [comment]
Hey @XoMEX that's great! I didn't really think of that. I'm using mitmproxy and can see regular https traffic, but the foyer api calls aren't going through because of ssl pinning. How did you bypass that?

@magicse thanks for the links. That `/baseconfig` file looks like a bunch of feature flags for rolling out gradually. And a list of all cast apps for some reason.

**Comment by XoMEX** [comment]
I don't think the api has pinned certs. The only pinned cert I found in the app was for the google home (only cert files in the app). But your device has to trust the certificate. I don't remember the exact way I got the certificate into android but it was with the help of some of these guides:
- https://docs.mitmproxy.org/stable/howto-install-system-trusted-ca-android/
- https://blog.nviso.eu/2018/01/31/using-a-custom-root-ca-with-burp-for-inspecting-android-n-traffic/
- https://github.com/oemunlock/burp_der_cert_to_android_cert
- https://awakened1712.github.io/hacking/hacking-install-ca-android/
- possibly google *android install /system/etc/security/cacerts/*

The gist is:
1. Generate a pen ca cert and key
2. Install the cert into android; Rename it as in the tutorials and copy it into that system folder
3. Tell mitmproxy to use the ca cert+key (by default it uses `~/.mitmproxy/mitmproxy-ca.pem`, I just copied it there)

Interestingly enough the cert I exported from burpsuite worked; I could not get the cert generated by mitmproxy to work (had no details in settings/trusted certs)
Interestinly 2: burpsuite did not work for me to capture the traffic. I guess mitm-ing tls is always annoying...

**Comment by rithvikvibhu** [comment]
I have already copied the cert to /system and it shows up in Trusted System Certs. The other apps seem to work fine. Which version of the GH app are you using?

And these are the only options for mitmproxy, right?
```sh
SSLKEYLOGFILE="$PWD/.mitmproxy/sslkeylogfile.txt" mitmproxy -p 8083
```

**Comment by XoMEX** [comment]
Version 2.22.1.11 on android 9 (x86_64). That command looks correct (assuming Linux; Windows has a weirder syntax for env flags. Assuming .mitmproxy exists and you have the correct port set as proxy)

**Comment by rithvikvibhu** [comment]
This is really weird. All other requests show up. Even the foyer api calls show up, but there's no response when the proxy is on. Requests to `clients3.google.com` work fine, protobuf response shows up fine.
![image](https://user-images.githubusercontent.com/5113343/84383419-df582380-ac09-11ea-8ae3-ff9c937c8057.png)

**Comment by XoMEX** [comment]
Yes, somehow mitmproxy does not capture the contents, I assume it has to do with http2. That is what I use Wireshark for. Just capture your ethernet/wifi and filter for `tls and http2`.

**Comment by rithvikvibhu** [comment]
But the requests don't even go through. The app doesn't show anything other than network issues. It doesn't show up in Wireshark (since the request isn't even made completely).

I'm gonna try a new VM and the latest app (on android 5, app v2.14 right now).

**Comment by XoMEX** [comment]
I just tried to create a cert using openssl and getting that into android. I did not get that working, so I guess I better not touch my working cert (>.<').
When using mitmproxy and opening chrome on Android, do https sites load? If yes, your proxy is working. Then you should be able to capture traffic using Wireshark and the key log file.
In mitmproxy the responses sometimes do not show up. As mentioned, I suspect this is due to http2.

**Comment by XoMEX** [comment]
I tried to get a reproducible way to generating and getting a cert into android and did not get it working. So I guess I have to keep using my working one I exported from burpsuite once. So I can't really help there. When I imported a newly generated cert in android it showed up in the settings, but when clicking on it, it did not have any details in it (Issuer, Subject, Serial number, Fingerprint, ...).

Are you sure the requests aren't completed? If you open chrome and browse to some https page, does it load? If yes, the request goes through. I think mitmproxy has problems recording fragmented responses in HTTP2. I also don't see responses in there very often. But Wireshark will capture them just fine, and using the keylogfile also decrypt them. (I set the tls version in mitmproxy to 1.2; 1.3/default might work too)
In your screenshot some requests have a 200 response code, thus the server answered, i.e. the tls handshake, request went through. Thus, the server can (and does) answer.

**Comment by AngeloD2022** [comment]
Are the request parameters needed? I can try, however, I know that google does certain things differently with iOS devices.  Is this used to authenticate with the local API for Google Smart Speakers?

**Comment by AngeloD2022** [comment]
One of the things that I know they do differently for iOS is that they do not SSL-pin their requests -or the majority of them, at least.

**Comment by AngeloD2022** [comment]
Also, knowing what specifically is this endpoint "getHomeGraph" is used for would be helpful

**Comment by rithvikvibhu** [comment]
@XoMEX yes, the https traffic shows up fine in both mitmproxy and wireshark (using the SSL logfile). Even other HTTPS requests from the GH app (like `https://clients3.google.com` and `https://android.googleapis.com`) work as expected.

It's only to `https://googlehomefoyer-pa.googleapis.com` are weird. They show up in mitmproxy so it is definitely being proxied. But there is no response. Even the app errors out.

Considering that googlehomefoyer-pa uses grpc, it's possible that mitmproxy cannot proxy them properly. I still don't get how the grpc showed up in your wireshark with no problem. Did you have to enable any protocol, setting, etc?

@AngeloD2022 I'm not sure here, but can grpc requests be replayed like regular http requests? I think they need special clients for this, curl can't handle it.

**Comment by XoMEX** [comment]
I don't remember doing any fancy settings, besides telling mitmproxy to use tls 1.2. I use Wireshark and mitmweb 5.1.1 on Windows. What I remember, is that the app did not like it to run the initial setup via the proxy. But once I was inside the overview it seemed to work.
Retrying it now, I see that not all requests are answered for some reason (possibly the client transmits some state in the request, and if that is not old enough, the cloud does not answer? Or the cloud is unstable?). Still I get an occasional response. I found them using `http2` as a filter and then using ctrl+f as in the screenshot, searching for `GetHomeGraph, Response`. Clicking "Find" highlighted one response
![image](https://user-images.githubusercontent.com/7779965/84531645-3252d900-ace5-11ea-83da-eef32f8bb1de.png)

**Comment by rithvikvibhu** [comment]
A few things I've found:

- It is indeed mitmproxy's problem, it can't proxy grpc properly https://github.com/mitmproxy/mitmproxy/issues/3052#issuecomment-381735135.
- I used https://github.com/bradleyjkemp/grpc-tools **and it worked!** Sure, had to create a CA signed cert for *.googleapis.com and use that. I've pasted what worked for me below.
- Curl cannot do grpc. Something like https://github.com/fullstorydev/grpcurl is needed, but it requires proto files, which we don't have. We could write one though (I think).

@XoMEX Huh, I missed that part from your previous comments. I'll try setting mitmproxy to tls 1.2 and try again. This will probably work imo

**Comment by rithvikvibhu** [comment]
Idk why, setting server tls version had no effect (mitmproxy docs say that tls1.2 is anyway the default). Anyway, grpc-tools worked for me, so yay. I'll try to script it to make this directly to the server now.

**Comment by rithvikvibhu** [comment]
Okay, progress.

I started with a simpler request `google.internal.home.foyer.v1.HomeControlService/GetAssistantRoutines` before starting with big ones like `getHomeGraph`.

Looking at the message format, I manually created a sample proto file. Since these values were mostly strings, it was easier to understand what they mean. This file could successfully be used with grpcurl.
I couldn't find foyer proto anywhere, so this is going to be based on guesses and manual referencing different people's data.

To make requests (first get [grpcurl](https://github.com/fullstorydev/grpcurl)):
```sh
# assume file exists at /path/to/protos/google/internal/home/foyer/v1.proto
./grpcurl -H 'authorization: Bearer ya29.a0Af****' -v -import-path /path/to/protos -proto /path/to/protos/google/internal/home/foyer/v1.proto googlehomefoyer-pa.googleapis.com:443 google.internal.home.foyer.v1.HomeControlService/GetAssistantRoutines
```

The only header needed (at least for this request) is the auth token. I used the one directly from the app, but thanks to @AngeloD2022 (who is working on an [OnHub Desktop app](https://github.com/AngeloD2022/OnHubDesktop)) there's already a script (as well as a webpage) to get this token easily. I haven't tested it yet, but it should work.

I'll keep updating these proto files as they are (sort of) ready here https://drive.google.com/drive/folders/1RvnN3y-G23pd2SWHmfV_7sef8QU5GNF4?usp=sharing
It only contains GetAssistantRoutines as of now.

**Comment by magicse** [comment]
Mat be this help (this about analyzing GRPC traffic )
Quote
Carl Mastrangelo | 09.02.18
Disclaimer: I work on gRPC so my opinion is quite biased.
Debugging is difficult from an external program for a few reasons:

1.  Traffic is usually encrypted.  HTTP/2 forces us to use ciphers that can't easy dump the TLS keys to disk for use with wireshark or other.  This is a security feature of ephemeral keys, but also a debugging hindrance.
2.  There are usually application/library level knobs that don't show up in the traffic.  For example  the socket option TCP_NOTSENT_LOWAT changes when writes take place, but is never visible from the traffic.    
3.  gGRPC has it's own protocol that requires headers and data frames to behave in a way a normal HTTP/2 traffic may not.  (for example, binary headers have to be base 64 decoded based on the header name).  

The most helpful tools I have used are dumping the http/2 frames directly from the program before they are encrypted and sent.   Secondly, running strace against the binary really help in telling why reads and writes may be slow, and what unusual socket operations are happening.  Lastly, dumping the HTTP/2 flow control window from the application itself is quite helpful for debugging slowdowns and stalls.   Only the changes in window are sent on the wire, but never the absolute amounts.  I guess what I'm getting at is that debugging from outside the program is much harder than having internal probes to the inside.


There is something really cool coming up called Channelz (search this forum for it).  This is a special gRPC service that you can enable to find out all the channels and servers you have running, what their channel state is, how active they are, and low level info about the sockets.  It will even have a channel trace expose that lets you inquire at runtime what interesting events have happened on a channel recently.  (For example, When re resolving DNS names, there was a failure that caused some of the subchannels to go idle).   Right now the spec is finalized, and the implementation is underway.   Eventually we will also release a client (probably JS), so you can view this debug into directly in your browser.

**Comment by magicse** [comment]
Also endpoint for homegraph API https://homegraph.googleapis.com/$discovery/rest?version=v1

**Comment by magicse** [comment]
Google home mini has 2 locales (as I saw in sources). One for chromecast part and other fro assistant part. If someone find how change locale for assistant not chromecats through GHlocalAPI please inform.

**Comment by rithvikvibhu** [comment]
@magicse I agree with most of it, but the problem is that, in this case, we don't have access to the server to enable extensions and the client (app) is obfuscated, debugging isn't easy.

That endpoint is for the public, which is `google.home.graph.v1.HomeGraphApiService`. We need `google.internal.home.foyer.v1.StructureService` to get the complete home graph.

**Comment by rithvikvibhu** [comment]
I've updated the proto file for getHomeGraph. My [previous comment](#issuecomment-643587959) has examples how to run it.

Now, this is only based on my home graph. Since I don't have a lot of devices, there might be some fields missing. I don't know what happens when the definition has missing fields. Please try it out and comment here if it works with your graphs or not.

**Comment by Leatherface75** [comment]
Works fine here

**Comment by XoMEX** [comment]
Works for me too. I have a home mini, a chromecast and two third-party vacuums. All show up. The chromecast and the home mini have a local auth token.

I also took a look at the bearer and where it is coming from:
`POST https://android.googleapis.com/auth`
Headers:
```
device:             <some id?>
app:                com.google.android.apps.chromecast.app
Accept-Encoding:    gzip
User-Agent:         GoogleAuth/1.4 (x86_64 PI); gzip
content-length:     719
content-type:       application/x-www-form-urlencoded
Host:               android.googleapis.com
Connection:         Keep-Alive
```
Post data
```
androidId:                    <some id, as in header>
lang:                         en-US
google_play_services_version: 201817032
sdk_version:                  28
device_country:               de
app:                          com.google.android.apps.chromecast.app
check_email:                  1
oauth2_foreground:            1
Email:                        <my email>
has_permission:               1
token_request_options:        <8 Alphanumerical chars>
service:                      oauth2:https://www.googleapis.com/auth/homegraph
client_sig:                   <40 hex chars/20B>
callerPkg:                    com.google.android.apps.chromecast.app
Token:                        <some token>
callerSig:                    <same as client_sig>
```
I haven't touched OAuth in a while, but I feel like the client sends too many fields for normal OAuth. It could be possible, that less data also works. Maybe the error when not supplying an auth token helps
> ERROR:
  Code: Unauthenticated
  Message: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.

**Comment by AngeloD2022** [comment]
I have seen a request similar to this when the app requests a URI for OAuth the consent screen. This was a more complex one to reverse, but I managed to make something working.
https://www.angelod.com/onhubauthtool

You can use the source code here: https://github.com/AngeloD2022/OnHubDesktop/blob/master/TemporaryAuthSolution/getToken.html

In order to make a request regarding the GH app, you must change some things. Copying the following from a Reddit conversation @rithvikvibhu and I had.

App Version: "2.23.112" (Not entirely sure this needs to be legitimate, but putting it here just in case...)

Package Name: "com.google.Chromecast" (Yes, this is the iOS app's bundle identifier 🙄)

Redirect URI: "com.google.sso.498579633514-hhlrn8mcjv1427j0s19dgfoe5cqaba4l:/authCallback"

Client ID: "498579633514-hhlrn8mcjv1427j0s19dgfoe5cqaba4l.apps.googleusercontent.com"

Mediator Client ID: "936475272427.apps.googleusercontent.com"

All of the rest is either can remain unchanged or is generated randomly–– and the funny thing is... IT WORKS!? 😂😂 Here's what the response looks like:

{ "advice_code": "EMBEDDED", "uri": "THE LOGIN FORM URI", "client_state": "RANDOM DATA", "iosguard_challenge": { "device_challenge": "RANDOM DATA" } }

**Comment by AngeloD2022** [comment]
@XoMEX And change the same for the oauthcode to refreshtoken request. (Inside of getToken())

**Comment by Drakulix** [comment]
> **Edit:** Turns out the tokens used by the onhub app are different from what we need. I wrote a python script that does it: https://gist.github.com/rithvikvibhu/952f83ea656c6782fbd0f1645059055d

@rithvikvibhu Under which license is your script?
I would love to update my home assistant component to make use of this generation method instead of extracting the graph file from an android device.

**Comment by Leatherface75** [comment]
Yeah that would be a big improvement and would make it possible to get it in as an official component again.

**Comment by rithvikvibhu** [comment]
@XoMEX yes, for the request (to get access tokens), only the `client_sig`, `service`, `app`, and `token` fields are required, others are optional.

There was already a python package that handled getting both kinds of tokens, so I didn't have to write manual requests.

@AngeloD2022 Huh, I thought that page was only for getting refresh tokens (`1/***`). I'll try it out soon. Is it possible to get access tokens (`ya29.***`) this way?

@Drakulix It's just a wrapper script with some values that calls functions from the `gpsoauth` package. Feel free to use it anywhere you want, gonna add an MIT license in a while.

**Comment by magicse** [comment]
Hi @rithvikvibhu, GetAssistantRoutines work well. But I cant get

**Comment by magicse** [comment]
Ah sorry I forgot change HomeControlService to StructuresService service

**Comment by luckydevil13** [comment]
How to dump (see grpc traffic?) any how to to setup mitm + grpcproxy ?

**Comment by rithvikvibhu** [comment]
@luckydevil13 to get the whole thing working, as of now, the only things needed are:
- An access token - [This gist](https://gist.github.com/rithvikvibhu/952f83ea656c6782fbd0f1645059055d) explains how to get one
- The proto file - [This drive folder](https://drive.google.com/drive/folders/1RvnN3y-G23pd2SWHmfV_7sef8QU5GNF4?usp=sharing) has it.
- grpcurl installed - [GitHub repo / releases](https://github.com/fullstorydev/grpcurl/)

How to actually make the request: [This comment](https://github.com/rithvikvibhu/GHLocalApi/issues/39#issuecomment-643587959)

However, if you're interested in actually looking into what's happening, then get [grpc-tools](https://github.com/bradleyjkemp/grpc-tools/) and [mitmproxy](https://docs.mitmproxy.org/stable/).
I don't see how `grpcproxy` would be useful in this case. Things you could do with these tools:
- Android proxied through `mitmproxy` to capture HTTP(s) requests and responses
- Android proxed through `grpcdump` to capture grpc requests and responses
- Wireshark over the VM's network adapter, can be used with either `mitmproxy` or `grpcdump`
- `grpcreplay` to replay dumped requests
- `grpcurl` to make any new grpc requests (need proto files though)

**Comment by unchartedshark** [comment]
Hey thanks for this, this has been great to follow along to.

I've been able to get the local auth token's for my devices but when I make a request via post man (Adding the Token as a Bearer Token) and making a request to /setup/assistant/alarms port 8008 return 403 forbidden and port 8443 doesn't return anything.

I'm assuming I'm doing something wrong here.

**Comment by magicse** [comment]
Hi @rithvikvibhu . I need some help. How did you create the proto (protos/google/internal/home/foyer/v1.proto) file? I'm trying to figure out how to transfer a configuration to a homegraf, 
For example, if I want to change the language locale via gRPC for home device.

**Comment by rithvikvibhu** [comment]
Honestly, It's manual work. First used `grpc-dump` to get a sample request/response in JSON. Then used https://www.site24x7.com/tools/json-to-protobuf.html to convert it to proto. Now this will give in v2, so there's again manual work to modify it into proto v3. The v2 definition gives a starting point, but there are still required to make it v3 compatible. This includes removing the "required" keyword from everywhere, setting the correct types, etc.

Because of the way the converter works, the names don't work. So you'll have to rename all the message names. Also, create a service and method definitions.

You can use the existing file (`protos/google/internal/home/foyer/v1`) as a reference and kinda duplicate the services section and rename it to the one you want.

**Comment by rithvikvibhu** [comment]
Hi everyone!

It's been exactly 1 year(!) (in 3 hours actually) since this issue was opened. Now that the API can be used again, I refreshed the website with a new design (finally mobile-friendly), features like search and updated detailed docs.

It's available at the same site: https://rithvikvibhu.github.io/GHLocalApi/

It's been great to be a part of an active project like this one.
Thanks to @Leatherface75 and @Drakulix for the scripts to make use of an Android VM and creating components that can be used by everyone (sadly, I never got to use this).
@magicse who revived this issue after ~8 months with the source dump (along with other info) and for testing everything posted here.
@XoMEX for examining the dump, finding how the local auth token works, and for the first capture of the grpc request for getting HomeGraph.
@AngeloD2022 for sending the decrypted iOS GH app and the OAuth auth {website, code, info}. It's a python script for now, but being able to do this with a website will be useful.

I'd like to thank everyone in this thread for posting new findings, testing scripts, and creating new libraries to make it easier to consume the API.

With 128 comments, this issue is getting long, so I'll be closing it. If there are any issues, please [create a new issue here](https://github.com/rithvikvibhu/GHLocalApi/issues/new)

**Comment by sjbeskur** [comment]
Hi @rithvikvibhu,
Your work on this has been excellent and very insightful.   In my attempts to understand how to connect and use the chromecast/ google home API I created a docker image to install prereqs and work through your examples.    The repo may be obtained here:  https://github.com/advanced-data-machines/chromeblast.   Per your excellent work, I am able to successfully obtain the  Master token then use this to get the Access Token, then (if the device is registered) use the Access Token to get the Local-Auth-Token (required by all of the POST actions).  

Even with the Local-Auth-Token set as the value for "cast-local-authentication-token" I remain  unable to issue successful queries to any rest methods that require the "cast-local-authorization-token" in the header?   I suspect I am missing something obvious but not sure what else to try.   Any help would be greatly appreciated.
